File: backend/src/main/java/com/billingapp/service/AuthService.java
package com.billingapp.service;

import com.billingapp.dto.auth.*;
import com.billingapp.entity.User;
import com.billingapp.exception.BadRequestException;
import com.billingapp.exception.ResourceNotFoundException;
import com.billingapp.repository.UserRepository;
import com.billingapp.security.JwtTokenProvider;
import com.billingapp.security.UserPrincipal;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider tokenProvider;
    private final AuthenticationManager authenticationManager;

    @Transactional
    public AuthResponse signup(SignupRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BadRequestException("Email is already registered");
        }

        User user = User.builder()
                .fullName(request.getFullName())
                .email(request.getEmail().toLowerCase())
                .password(passwordEncoder.encode(request.getPassword()))
                .phoneNumber(request.getPhoneNumber())
                .provider(User.AuthProvider.LOCAL)
                .emailVerified(false)
                .build();

        user = userRepository.save(user);
        log.info("User registered successfully: {}", user.getEmail());

        return generateAuthResponse(user);
    }

    @Transactional
    public AuthResponse login(LoginRequest request) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        request.getEmail().toLowerCase(),
                        request.getPassword()
                )
        );

        SecurityContextHolder.getContext().setAuthentication(authentication);
        UserPrincipal userPrincipal = (UserPrincipal) authentication.getPrincipal();

        User user = userRepository.findById(userPrincipal.getId())
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userPrincipal.getId()));

        return generateAuthResponse(user);
    }

    @Transactional
    public AuthResponse refreshToken(RefreshTokenRequest request) {
        String refreshToken = request.getRefreshToken();

        if (!tokenProvider.validateToken(refreshToken)) {
            throw new BadRequestException("Invalid refresh token");
        }

        Long userId = tokenProvider.getUserIdFromToken(refreshToken);
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));

        if (!refreshToken.equals(user.getRefreshToken())) {
            throw new BadRequestException("Refresh token mismatch");
        }

        return generateAuthResponse(user);
    }

    @Transactional
    public void logout(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));
        user.setRefreshToken(null);
        userRepository.save(user);
    }

    private AuthResponse generateAuthResponse(User user) {
        String accessToken = tokenProvider.generateAccessToken(user.getId());
        String refreshToken = tokenProvider.generateRefreshToken(user.getId());

        // Store refresh token
        user.setRefreshToken(refreshToken);
        userRepository.save(user);

        boolean hasShop = user.getShop() != null;
        Long shopId = hasShop ? user.getShop().getId() : null;

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .tokenType("Bearer")
                .expiresIn(tokenProvider.getAccessTokenExpirationMs())
                .user(AuthResponse.UserInfo.builder()
                        .id(user.getId())
                        .fullName(user.getFullName())
                        .email(user.getEmail())
                        .phoneNumber(user.getPhoneNumber())
                        .profileImageUrl(user.getProfileImageUrl())
                        .hasShop(hasShop)
                        .shopId(shopId)
                        .build())
                .build();
    }
}

File: lib/core/services/auth_service.dart
import 'package:billing_app/core/services/api_service.dart';
import 'package:billing_app/core/services/storage_service.dart';
import 'package:billing_app/models/auth_models.dart';

class AuthService {
  final ApiService _api = ApiService();
  final StorageService _storage = StorageService();

  Future<ApiResponse<AuthResponse>> signup({
    required String fullName,
    required String email,
    required String password,
    String? phoneNumber,
  }) async {
    final response = await _api.post<AuthResponse>(
      '/v1/auth/signup',
      body: {
        'fullName': fullName,
        'email': email,
        'password': password,
        if (phoneNumber != null) 'phoneNumber': phoneNumber,
      },
      requiresAuth: false,
      fromJson: (json) => AuthResponse.fromJson(json),
    );

    if (response.isSuccess && response.data != null) {
      await _saveAuthData(response.data!);
    }

    return response;
  }

  Future<ApiResponse<AuthResponse>> login({
    required String email,
    required String password,
  }) async {
    final response = await _api.post<AuthResponse>(
      '/v1/auth/login',
      body: {
        'email': email,
        'password': password,
      },
      requiresAuth: false,
      fromJson: (json) => AuthResponse.fromJson(json),
    );

    if (response.isSuccess && response.data != null) {
      await _saveAuthData(response.data!);
    }

    return response;
  }

  Future<ApiResponse<AuthResponse>> refreshToken() async {
    final refreshToken = await _storage.getRefreshToken();
    if (refreshToken == null) {
      return ApiResponse.error('No refresh token available');
    }

    final response = await _api.post<AuthResponse>(
      '/v1/auth/refresh',
      body: {'refreshToken': refreshToken},
      requiresAuth: false,
      fromJson: (json) => AuthResponse.fromJson(json),
    );

    if (response.isSuccess && response.data != null) {
      await _saveAuthData(response.data!);
    }

    return response;
  }

  Future<void> logout() async {
    try {
      await _api.post('/v1/auth/logout');
    } finally {
      await _storage.clearAll();
    }
  }

  Future<void> _saveAuthData(AuthResponse authResponse) async {
    await _storage.saveTokens(
      accessToken: authResponse.accessToken,
      refreshToken: authResponse.refreshToken,
    );

    if (authResponse.user != null) {
      await _storage.saveUserData(
        userId: authResponse.user!.id,
        email: authResponse.user!.email,
        hasShop: authResponse.user!.hasShop,
        shopId: authResponse.user!.shopId,
      );
    }
  }

  Future<bool> isLoggedIn() => _storage.isLoggedIn();
  
  Future<bool> hasShop() => _storage.hasShop();
}
